video_calls: Add Webex as a Video Call Provider.

GregoryDRowe opened the pull request:
This adds new server endpoints for connecting to Webex
via OAuth as wells as creating a new Webex meeting.
Refactored existing Zoom OAuth implementation to
reduce code duplication.
Updated frontend to include settings for Webex video calls
and UI to create messages for generated Webex meetings.

**Testing plan:** New front and backend tests written and passing.
Manual testing performed for Webex and Zoom calls.

timabbott commented on zproject/default_settings.py:
```diff
@@ -90,9 +90,21 @@
+VIDEO_WEBEX_CLIENT_ID = get_secret('video_webex_client_id', development_only=True)
+VIDEO_WEBEX_CLIENT_SECRET = get_secret('video_webex_client_secret')
+# Length of alphanumeric password generated when a Webex meeting is created
+VIDEO_WEBEX_PASSWORD_LENGTH = 11
```
How did you pick the number 11 here?

GregoryDRowe replied:
From this document: https://help.webex.com/en-us/so30qz/Required-Password-Enforcement-for-Webex-Meetings-Events-and-Training-Sessions#:~:text=Webex%20sites%20will%20be%20configured,and%20to%20slow%20channel%20sites specifically 'The default passwords applied will be 11 or more characters, including mixed case, numbers, and special characters.'

timabbott commented on zproject/default_settings.py:
```diff
@@ -90,9 +90,21 @@
+# Length of Webex Meeting in minutes
+VIDEO_WEBEX_MEETING_LENGTH = 60
+VIDEO_WEBEX_AUTO_RECORD_MEETING = False
+VIDEO_WEBEX_ANY_USER_CAN_COHOST = False
```
I think this is OK for an initial version, but longer-term I imagine we'll want to somehow adjust the UI to allow the user to specify these settings at meeting creation time (E.g. maybe the video call button should pop up a modal to make these selections).  Maybe worth adding a comment noting that plan.

timabbott commented on zproject/urls.py:
```diff
@@ -716,6 +722,10 @@
 
+    # Used to generate a Webex video call URL
+    path('calls/webex/register', register_webex_user),
+    path('calls/webex/complete', complete_webex_user),
```
Maybe put this above the BigBlueButton endpoints since it has a more similar model to Zoom?

timabbott commented on zerver/tests/test_create_video_call.py:
```diff
@@ -129,18 +139,18 @@ def test_create_video_request_error(self) -> None:
     @responses.activate
     def test_deauthorize_zoom_user(self) -> None:
-        responses.add(responses.POST, "https://api.zoom.us/oauth/data/compliance")
+        responses.add(responses.POST, "https://zoom.example.com/oauth/data/compliance")
```
Generally, we aim to do this sort of refactor (along with the settings and structural changes to the Zoom integration) as a series of independent, atomic, preparatory commits, so that we can easily review and merge those refactors that are relatively fast to review while we do any back-and-forth on the implementation for the new Webex feature.  This ends up saving a lot of reviewer time for any PR that needs multiple review round trips.

Check out the Zulip commit guidelines for more details: https://zulip.readthedocs.io/en/latest/contributing/version-control.html

GregoryDRowe replied:
Hi Tim, I've broken this out into 7 commits now: Refactor these tests, refactor the zoom_xhrs method, refactor the zoom API into a generic base, and then 4 stages of adding the new Webex stuff (token, api, frontend, documentation).

timabbott commented on zerver/models.py:
```diff
@@ -1116,6 +1123,7 @@ class UserProfile(AbstractBaseUser, PermissionsMixin):
     onboarding_steps: str = models.TextField(default='[]')
 
     zoom_token: Optional[object] = JSONField(default=None, null=True)
+    webex_token: Optional[object] = JSONField(default=None, null=True)
```
@andersk what are your thoughts on the data model here -- should we be trying to refactor this to use the same field for both providers in some way?

timabbott commented on static/js/compose.js:
```diff
@@ -886,6 +890,28 @@ exports.needs_subscribe_warning = function (user_id, stream_id) {
+    );
+
+    insert_video_call_url(details.url, target_textarea);
+}
```
Can you post a screenshot of what the resulting message looks like?  I'm curious if we should be using bulleted lists or the like here.

GregoryDRowe replied:
![ZulipMeeting](https://user-images.githubusercontent.com/65816804/97283715-14bb8180-1838-11eb-8a1c-d91b248217c0.PNG)

timabbott commented on static/js/compose.js:
```diff
@@ -184,13 +185,16 @@ exports.abort_xhr = function () {
+    exports.webex_token_callbacks.delete(key);
+    if (exports.video_call_xhrs.has(key)) {
+        exports.video_call_xhrs.get(key).abort();
+        exports.video_call_xhrs.delete(key);
```
This refactor to the `video_call_xhrs` variable name would be a great prep refactoring commit.

timabbott commented:
@GregoryDRowe huge thanks for writing this!  Overall this looks very thoughtfully done in the style of the existing codebase, which I'm thankful for.  The plumbing looks great; the one section I haven't look at closely is the `zerver/views/video_calls.py` refactor.  @andersk can you take point on reviewing that, as its original author?

timabbott commented on templates/zerver/help/start-a-call.md:
```diff
@@ -77,3 +94,4 @@ setting the provider to "None".
 
 [big-blue-button-configuration]: https://zulip.readthedocs.io/en/latest/production/video-calls.html#big-blue-button
 [zoom-configuration]: https://zulip.readthedocs.io/en/latest/production/video-calls.html#zoom
+[webex-configuration]: https://zulip.readthedocs.io/en/latest/production/video-calls.html#webex
```
I think you forgot to add the documentation in `docs/production/video-calls.md` that this links to.

GregoryDRowe replied:
Thanks, I have added it now.

timabbott commented:
The changes you made all look good; feel free to squash them into the original commit (before potentially splitting out commits for the preparatory refactorings)

GregoryDRowe committed:
video_calls: Refactor Zoom API to support new OAuth video applications.

Refactor contents of the 4 Zoom API endpoints
to an abstract base class in order to support
addition of other OAuth backed video clients.
Refactor Zoom API URL into settings to ease
testing and support other Zoom based instances
ie. ZoomGov, etc.

GregoryDRowe committed:
video_calls: Add webex_token to User_Profile.

Add webex_token to the user profiles in preparation
for adding the ability for the user to generate Webex
meetings. This property will store the OAuth token
provided by Webex once the user has approved linking
their Webex and Zulip accounts.
Add has_webex_token action for when a Webex token is
added or removed from a user.
Refactor the clear_zoom_token_on_logout signal to
clear_video_call_tokens_on_logout and made it clear
the webex_token as well as the zoom_token on user
logout.
Add do_set_webex_token for updating the users
webex_token

GregoryDRowe committed:
video_calls: Add new API endpoints for generating a Webex meeting.

Add calls/webex/register and calls/webex/complete API endpoints
for acquiring a Webex token via OAuth and calls/webex/create to
generate a new Webex meeting and return the details.
Add default Webex meeting parameters to the settings.

GregoryDRowe committed:
video_calls: Add Webex as a new Video Chat Provider.

Add Webex as a new Video Chat Provider.
Add event handling to the front end for the
has_webex_token server event.
Add handling for Webex as a video chat provider to
compose.

GregoryDRowe committed:
video_calls: Add documentation for Webex as a Video Chat Provider.

timabbott commented:
I merged the first two commits and pushed back here; thanks for splitting those.  For the next commit, what's your thinking on changing the test settings to do `zoom.us` => `zoom.example.com` for the URLs computed?  I guess that probably is safer in case the tests end up doing network requests by accident, but I think we block all outgoing HTTP requests in the tests anyway.

GregoryDRowe commented:
> I merged the first two commits and pushed back here; thanks for splitting those. For the next commit, what's your thinking on changing the test settings to do `zoom.us` => `zoom.example.com` for the URLs computed? I guess that probably is safer in case the tests end up doing network requests by accident, but I think we block all outgoing HTTP requests in the tests anyway.

My idea was to test that everywhere was using the new setting, rather than the old hard coded URL. Changing this to a setting had the double advantage of making it much easier to manually test locally and allows for other Zoom hostings to be used (such as ZoomGov).

timabbott commented:
I was not familiar with ZoomGov, that's certainly a helpful argument in favor of making the URL configurable.

GregoryDRowe commented:
Hi @timabbott, Sorry to bug you, but is there anything else that I need to do to progress this PR?

laurynmm commented:
Webex was added as a supported video call provider for self-hosted Zulip servers in #37540. @GregoryDRowe - Thanks for your work on this!
